blacklistfilter/ipblacklistfilter.cpp

```diff
@@ -0,0 +1,740 @@
+            trap_send(0, detection, ur_rec_fixlen_size(tmpl_det));
+            DBG((stderr, "IP detected on blacklist\n", ))
+        }
+
+        if (BL_RELOAD_FLAG) {
+            DBG((stderr, "Reloading blacklists\n"));
+            retval = reload_blacklists(v4_list, v6_list, bl_file);
+            if (retval == BLIST_FILE_ERROR) {
+                std::cerr << "ERROR: Unable to load update blacklist. Will use the old one instead." << std::endl;
+                continue;
+            }
+
+            // this lazy locking is fine, we don't need to reload the blacklists immediately
+            // and locking the mutex in every iteration is ineffective
+            pthread_mutex_lock(&BLD_SYNC_MUTEX);
+            BL_RELOAD_FLAG = 0;
+            pthread_mutex_unlock(&BLD_SYNC_MUTEX);
+        }
+
+        // TODO: Do we need any reconfiguration at all?
+        if (RECONF_FLAG) {
+            DBG((stderr, "Reconfiguration..\n"))
+
+            v4_list.clear();
+            v6_list.clear();
+
+            if (loadConfiguration((char*)MODULE_CONFIG_PATTERN_STRING, userFile, &config, CONF_PATTERN_STRING)) {
+                std::cerr << "Error: Could not parse XML configuration." << std::endl;
+                ur_free_template(templ);
+                ur_free_template(tmpl_det);
+                trap_finalize();
+                FREE_MODULE_INFO_STRUCT(MODULE_BASIC_INFO, MODULE_PARAMS)
+                return EXIT_FAILURE;
+            }
+
+            if (strcmp(config.watch_blacklists, "true") == 0) {
+                WATCH_BLACKLISTS_FLAG = true;
+            } else {
+                WATCH_BLACKLISTS_FLAG = false;
+            }
+
+            bl_file = config.blacklist_file;
+
+            // Load ip addresses from sources
+            retval = reload_blacklists(v4_list, v6_list, bl_file);
+
+            // If update from file could not be processed, return error
+            if (retval == BLIST_FILE_ERROR) {
+                fprintf(stderr, "Error: Unable to read bl_file '%s'\n", bl_file.c_str());
+                ur_free_template(templ);
+                ur_free_template(tmpl_det);
+                trap_finalize();
+                FREE_MODULE_INFO_STRUCT(MODULE_BASIC_INFO, MODULE_PARAMS)
+                return EXIT_FAILURE;
+            }
+
+            RECONF_FLAG = 0;
+        }
+
+    }
+
+    // Module is stopping, set appropriate flag
+    stop = 1;
+
+    // If set, send terminating message to modules on output
+    if (send_terminating_unirec) {
+        trap_send(0, "TERMINATE", 1);
+    }
+
+    // Clean up before termination
+    if (detection != NULL) {
+        ur_free_record(detection);
+        detection = NULL;
+    }
+
+    ur_free_template(templ);
+    ur_free_template(tmpl_det);
+
+    TRAP_DEFAULT_FINALIZATION();
+    FREE_MODULE_INFO_STRUCT(MODULE_BASIC_INFO, MODULE_PARAMS)
+
+    if (WATCH_BLACKLISTS_FLAG)
+        pthread_cancel(watcher_thread);
+
+    return EXIT_SUCCESS;
```

Comment:
missing trap_finalize()

Reply:
trap_finalize() is not necessary, since it is included in the macro TRAP_DEFAULT_FINALIZATION().